### ISSUE TITLE
Provider gets passed wrong source

### ISSUE THREAD

**Issue by KoenigsKind** [modelmapper/modelmapper#573]
### Problem
In our software we use a custom provider, to create the destination object depending on the source object and the destination class. Most of the time this works fine, but there are some edge cases, where it fails, as we receive the wrong source object in the provider.

The example below produces the following output, showing the class of the source object and the destination class:
> A → B
> AAA → BB
> AAA → BBB
> Integer → String
> 42

The problem is here the second line, in our understanding this should read `AA → BB`. For some reason it receives the child object as source.

I tested it with version 2.3.4 (currently on our production server) and version 2.3.9. Both showing the same problem.

### Example
This is a groovy script to demonstrate the problem:
```groovy
import org.modelmapper.ModelMapper
import org.modelmapper.Provider
import org.modelmapper.convention.MatchingStrategies


ModelMapper mapper = new ModelMapper()
mapper.configuration.matchingStrategy = MatchingStrategies.STRICT
mapper.configuration.provider = new MyProvider()

A input = new A(x: new AA(y: new AAA(z: 42)))
B output = mapper.map(input, B)

println output?.x?.y?.z


// ------- Classes ----------
class A {AA x}
class AA {AAA y}
class AAA {Integer z}

class B {BB x}
class BB {BBB y}
class BBB {String z}

class MyProvider implements Provider {
    @Override
    Object get(Provider.ProvisionRequest request) {
        println "${request.source?.class?.simpleName} → ${request.requestedType?.simpleName}"
    }
}
```

**Comment by chhsiao90** [comment]
Thanks for reporting the issue, fixed by #573

**Comment by chhsiao90** [comment]
ModelMapper v2.4.1 was just released including the fix.
